File: solutions/go/food-chain/food_chain.go
package foodchain

import (
	"fmt"
)

type Part = struct {
	name, comment, end string
	rolls              bool
}

var parts = []Part{
	{"fly", "I don't know why she swallowed the fly. Perhaps she'll die.", ".\nI don't know why she swallowed the fly. Perhaps she'll die.", false},
	{"spider", "It wriggled and jiggled and tickled inside her.", " that wriggled and jiggled and tickled inside her.", true},
	{"bird", "How absurd to swallow a bird!", ".", true},
	{"cat", "Imagine that, to swallow a cat!", ".", true},
	{"dog", "What a hog, to swallow a dog!", ".", true},
	{"goat", "Just opened her throat and swallowed a goat!", ".", true},
	{"cow", "I don't know how she swallowed a cow!", ".", true},
	{"horse", "She's dead, of course!", ".", false},
}

func preys(v int) string {
	part := parts[v-1]
	if !part.rolls {
		return ""
	}
	prev := parts[v-2]
	return fmt.Sprintf("\nShe swallowed the %s to catch the %s%s%s", part.name, prev.name, prev.end, preys(v-1))
}

func Verse(v int) string {
	part := parts[v-1]
	return fmt.Sprintf("I know an old lady who swallowed a %s.\n%s%s", part.name, part.comment, preys(v))
}

func Verses(start, end int) string {
	s := ""
	for i := start; i <= end; i++ {
		if s != "" {
			s += "\n\n"
		}
		s += Verse(i)
	}
	return s
}

func Song() string {
	return Verses(1, 8)
}

File: solutions/ruby/food-chain/food_chain.rb
class FoodChain
    module Animals
        class Animal
            def verse = ([swallow] + preys).join("\n") + "\n"
            def prey_on = Animals::ANIMALS.filter{_1.order < self.order}
            def name = self.class.name.split('::').last.downcase
            def swallow = "I know an old lady who swallowed a #{name}.\n#{comment}"
            def preys
                prev = prey_on.last
                if prev
                    ["She swallowed the #{name} to catch the #{prev.name}#{prev.end}"] + prev.preys
                else
                    []
                end
            end
            def end = "."
        end

        class Fly < Animal
            def order = 0
            def comment = "I don't know why she swallowed the fly. Perhaps she'll die."
            def end = ".\n#{comment}"
        end
        class Spider < Animal
            def order = 1
            def comment = "It wriggled and jiggled and tickled inside her."
            def end = " that wriggled and jiggled and tickled inside her."
        end
        class Bird < Animal
            def order = 2
            def comment = "How absurd to swallow a bird!"
        end
        class Cat < Animal
            def order = 3
            def comment = "Imagine that, to swallow a cat!"
        end
        class Dog < Animal
            def order = 4
            def comment = "What a hog, to swallow a dog!"
        end
        class Goat < Animal
            def order = 5
            def comment = "Just opened her throat and swallowed a goat!"
        end
        class Cow < Animal
            def order = 6
            def comment = "I don't know how she swallowed a cow!"
        end
        class Horse < Animal
            def order = 7
            def comment = "She's dead, of course!"
            def prey_on = []
        end
        ANIMALS = Animal.subclasses.map(&:new).sort{_1.order}
    end

    def self.song
        Animals::ANIMALS.map(&:verse).join("\n")
    end
end